ecosystem/indexer-grpc/indexer-grpc-table-info/src/table_info_parser.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Error;
+use aptos_api::context::Context;
+use aptos_api_types::TransactionOnChainData;
+use aptos_logger::{error, info, sample, sample::SampleRate};
+use aptos_sdk::bcs;
+use aptos_storage_interface::DbWriter;
+use aptos_types::write_set::WriteSet;
+use std::sync::Arc;
+use tonic::Status;
+
+type EndVersion = u64;
+
+pub struct TableInfoParser {
+    pub current_version: u64,
+    pub parser_task_count: u16,
+    pub parser_batch_size: u16,
+    pub highest_known_version: u64,
+    pub context: Arc<Context>,
+}
+
+pub struct TransactionBatchInfo {
+    pub start_version: u64,
+    pub num_transactions_to_fetch: u16,
+}
+
+impl TableInfoParser {
+    pub fn new(
+        context: Arc<Context>,
+        request_start_version: u64,
+        parser_task_count: u16,
+        parser_batch_size: u16,
+    ) -> Self {
+        Self {
+            current_version: request_start_version,
+            parser_task_count,
+            parser_batch_size,
+            highest_known_version: 0,
+            context,
+        }
+    }
+
+    pub async fn process_next_batch(
+        &mut self,
+        db_writer: Arc<dyn DbWriter>,
+    ) -> Vec<Result<EndVersion, Status>> {
+        let mut tasks = vec![];
+        let batches = self.get_batches().await;
+        let db_writer = db_writer.clone();
+
+        for batch in batches {
+            let db_writer = db_writer.clone();
+            let context = self.context.clone();
+            let ledger_version = self.highest_known_version;
+            let task = tokio::spawn(async move {
+                let raw_txns =
+                    Self::fetch_raw_txns_with_retries(context.clone(), ledger_version, batch).await;
```

Comment:
can we measure the duration of the fetch as well so we know the breakdown of duration for processing a whole batch?